### ISSUE TITLE
overfitting in this model？

### ISSUE THREAD

**Issue by ninesky110** [pdoublerainbow/bisenet-tensorflow#3]

**Comment by pdoublerainbow** [comment]
If you just test it in your own dataset, that must have question here, for the reason of domain shift. I did not train the code too hard, if you want train the model more better, I recommend you to add hard example mining code in training step, like this:
        # Hard example mining or not
        if top_k_percent_pixels == 1.0:
            # Compute the loss for all pixels.
            loss = tf.reduce_mean(loss)
        else:

            loss = tf.reshape(loss, [-1])
            num_pixels = tf.to_float(tf.shape(loss)[0])
            # Parameter prepare
            t_total = int(
                self.data_config['epoch'] * self.data_config['num_examples_per_epoch'] / self.data_config['batch_size'])
            hard_example_mining_step = get(self.data_config, 'hard_example_mining_step', int(t_total * 2 / 3))
            # Compute the top_k_percent pixels based on current training step.
            if hard_example_mining_step == 0:
                # Directly focus on the top_k pixels.
                top_k_pixels = tf.to_int32(top_k_percent_pixels * num_pixels)
            else:
                # Gradually reduce the mining percent to top_k_percent_pixels.
                global_step = tf.to_float(self.global_step)
                ratio = tf.minimum(1.0, global_step / hard_example_mining_step)
                top_k_pixels = tf.to_int32(
                    (ratio * top_k_percent_pixels + (1.0 - ratio)) * num_pixels)
            top_k_losses, _ = tf.nn.top_k(loss,
                                          k=top_k_pixels,
                                          sorted=True,
                                          name='top_k_percent_pixels')
            loss = tf.reduce_sum(top_k_losses)
            num_present = tf.reduce_sum(
                tf.to_float(tf.not_equal(top_k_losses, 0.0)))
            loss = self._div_maybe_zero(loss, num_present)

        tf.losses.add_loss(loss)

**Comment by cena001plus** [comment]
程序可读性不好

**Comment by pdoublerainbow** [comment]
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright @ 2019 Liming Liu     HuNan University
#
# I want to set this model as baseline which use residual mobile_net v2
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
import tensorflow as tf
from tensorflow.contrib import slim, metrics
from Dataset.cityscape import Cityscape
from models.model import Model
from backbone.mobilenet_v2 import mobilenet_v2
from utils.misc_utils import get
from utils.loss_util import lovasz_softmax


class Baseline(Model):
    def __init__(self, model_config, train_config, num_classes, mode):

        super(Baseline, self).__init__(model_config, train_config, num_classes, mode)
        self.net = None
        self.sup1 = None
        self.sup2 = None
        self.loss = None
        self.total_loss = None
        self.response = None
        self.global_step = None
        self.init_fn = None
        self.accuracy = None
        self.meanIOU = None
        self.recall = None
        self.precision = None

    def build_baseline(self, reuse=False):
        """
        Builds the baseline model.

        Arguments:
          reuse: Reuse variable or not

        Returns:
          Baseline model
        """
        batch_norm_params = self.model_config['batch_norm_params']
        init_method = self.model_config['conv_config']['init_method']
        weight_decay = self.model_config['conv_config']['weight_decay']

        if init_method == 'kaiming_normal':
            initializer = slim.variance_scaling_initializer(factor=2.0, mode='FAN_IN', uniform=False)
        else:
            initializer = slim.xavier_initializer()

        with tf.variable_scope('Baseline', reuse=reuse):
            # Build Bsaeline
            with slim.arg_scope([slim.conv2d, slim.separable_conv2d], weights_initializer=initializer,
                                normalizer_fn=slim.batch_norm, biases_initializer=None), \
                 slim.arg_scope([slim.batch_norm], is_training=self.is_training(), **batch_norm_params),\
                 slim.arg_scope([slim.conv2d], weights_regularizer=slim.l2_regularizer(weight_decay)),\
                 slim.arg_scope([slim.separable_conv2d], weights_regularizer=None, activation_fn=None):

                end_points = mobilenet_v2(self.images, expansion=6, depth_multiplier=1)

                # tail part
                # Take this as baseline, consider to change all conv to separable_conv,
                # or take asymmetric_conv into consider, but first use this as baseline,
                # After upsampling consider to use dilation separable conv?
                # All the place can change i leave a mark
                size = tf.shape(end_points['pool5'])[1:3]
                global_context = tf.reduce_mean(end_points['pool5'], [1, 2], keep_dims=True)
                global_context = slim.conv2d(global_context, 128, 1, [1, 1])
                global_context = tf.image.resize_bilinear(global_context, size=size)

                net_5 = slim.conv2d(end_points['pool5'], 128, [3, 3])  #
                net_4 = slim.conv2d(end_points['pool4'], 128, [3, 3])  #
                net_3 = slim.conv2d(end_points['pool3'], 128, [3, 3])  #

                net_5 = tf.add(net_5, global_context)
                net_5 = self.Upsampling(net_5, scale=2)
                net_5 = slim.conv2d(net_5, 128, kernel_size=[3, 3])    #
                net_4 = tf.add(net_4, net_5)
                net_4 = self.Upsampling(net_4, scale=2)
                net_4 = slim.conv2d(net_4, 128, kernel_size=[3, 3])    #

                net = tf.add(net_4, net_3)

                net_5 = slim.conv2d(net_5, 128, kernel_size=[3, 3])    #
                net_4 = slim.conv2d(net_4, 128, kernel_size=[3, 3])    #
                net = slim.conv2d(net, 64, kernel_size=[3, 3])

                net = slim.conv2d(net, self.num_classes, [1, 1], normalizer_fn=None, activation_fn=None, scope='logits')
                self.net = self.Upsampling(net, scale=8)

                if self.mode in ['train', 'validation', 'test', 'fine_tune']:
                    sup1 = slim.conv2d(net_5, self.num_classes, [1, 1], normalizer_fn=None, activation_fn=None,
                                       scope='supl1')
                    sup2 = slim.conv2d(net_4, self.num_classes, [1, 1], normalizer_fn=None, activation_fn=None,
                                       scope='supl2')
                    self.sup1 = self.Upsampling(sup1, scale=16)
                    self.sup2 = self.Upsampling(sup2, scale=8)

    def build_loss(self, weight=None):
        # Parameter prepare
        top_k_percent_pixels = get(self.data_config, 'top_k_percent_pixels', 1.0)

        loss1 = tf.nn.softmax_cross_entropy_with_logits(logits=self.net, labels=self.labels)
        loss2 = tf.nn.softmax_cross_entropy_with_logits(logits=self.sup1, labels=self.labels)
        loss3 = tf.nn.softmax_cross_entropy_with_logits(logits=self.sup2, labels=self.labels)

        if weight is not None:
            weights = self.labels * weight
            weights = tf.reduce_sum(weights, 3)
            loss1 = loss1*weights
            loss2 = loss2*weights
            loss3 = loss3*weights
        loss = loss1 + loss2 + loss3

        # Hard example mining or not
        if top_k_percent_pixels == 1.0:
            # Compute the loss for all pixels.
            loss = tf.reduce_mean(loss)
        else:

            loss = tf.reshape(loss, [-1])
            num_pixels = tf.to_float(tf.shape(loss)[0])
            # Parameter prepare
            t_total = int(
                self.data_config['epoch'] * self.data_config['num_examples_per_epoch'] / self.data_config['batch_size'])
            hard_example_mining_step = get(self.data_config, 'hard_example_mining_step', int(t_total * 2 / 3))
            # Compute the top_k_percent pixels based on current training step.
            if hard_example_mining_step == 0:
                # Directly focus on the top_k pixels.
                top_k_pixels = tf.to_int32(top_k_percent_pixels * num_pixels)
            else:
                # Gradually reduce the mining percent to top_k_percent_pixels.
                global_step = tf.to_float(self.global_step)
                ratio = tf.minimum(1.0, global_step / hard_example_mining_step)
                top_k_pixels = tf.to_int32(
                    (ratio * top_k_percent_pixels + (1.0 - ratio)) * num_pixels)
            top_k_losses, _ = tf.nn.top_k(loss,
                                          k=top_k_pixels,
                                          sorted=True,
                                          name='top_k_percent_pixels')
            loss = tf.reduce_sum(top_k_losses)
            num_present = tf.reduce_sum(
                tf.to_float(tf.not_equal(top_k_losses, 0.0)))
            loss = self._div_maybe_zero(loss, num_present)

        tf.losses.add_loss(loss)

        self.loss = tf.reduce_mean(loss1)
        self.total_loss = tf.losses.get_total_loss()

        # Tensorboard inspection
        colors = Cityscape.get_class_colors()
        shape = tf.shape(self.labels)
        img_mean = self.data_config['img_mean']
        tf.summary.image('image', self.images + img_mean, family=self.mode, max_outputs=1)
        tf.summary.image('GT', tf.reshape(tf.matmul(tf.reshape(self.labels, [-1, self.num_classes]), colors),
                                          [-1, shape[1], shape[2], 3]), family=self.mode, max_outputs=1)
        tf.summary.image('response', tf.reshape(tf.matmul(
            tf.reshape(tf.one_hot(tf.argmax(self.net, -1), self.num_classes), [-1, self.num_classes]), colors),
            [-1, shape[1], shape[2], 3]), family=self.mode, max_outputs=1)
        tf.summary.scalar('total_loss', self.total_loss, family=self.mode)
        tf.summary.scalar('loss', self.loss, family=self.mode)

        accuracy, accuracy_update = metrics.streaming_accuracy(tf.argmax(self.net, -1),
                                                               tf.argmax(self.labels, -1))
        mean_IOU, mean_IOU_update = metrics.streaming_mean_iou(predictions=tf.argmax(self.net, -1),
                                                               labels=tf.argmax(self.labels, -1),
                                                               num_classes=self.num_classes)
        recall, recall_update = metrics.streaming_recall(tf.argmax(self.net, -1), tf.argmax(self.labels, -1))
        precision, precision_update = metrics.streaming_precision(tf.argmax(self.net, -1), tf.argmax(self.labels, -1))
        with tf.control_dependencies([accuracy_update, mean_IOU_update, recall_update, precision_update]):
            tf.summary.scalar('accuracy', accuracy, family=self.mode)
            tf.summary.scalar('mean_IOU', mean_IOU, family=self.mode)
            tf.summary.scalar('recall', recall, family=self.mode)
            tf.summary.scalar('precision', precision, family=self.mode)
            accuracy = tf.identity(accuracy)
            mean_IOU = tf.identity(mean_IOU)
            recall = tf.identity(recall)
            precision = tf.identity(precision)
        self.accuracy = accuracy
        self.meanIOU = mean_IOU
        self.recall = recall
        self.precision = precision

    def lovasz_loss(self):
        loss = lovasz_softmax(self.net, self.labels)
        # loss2 = lovasz_softmax(self.sup2, self.labels)
        # loss3 = lovasz_softmax(self.sup1, self.labels)
        # loss = loss1
        tf.losses.add_loss(loss)
        self.loss = loss
        self.total_loss = tf.losses.get_total_loss()

    def predict(self):
        self.response = tf.argmax(self.net, -1)

    def build(self, reuse=False):
        """Creates all ops for training and evaluation"""
        with tf.name_scope(self.mode):
            self.build_inputs()
            self.build_baseline(reuse=reuse)

            if self.is_training():
                self.setup_global_step()

            if self.mode in ['train', 'validation', 'test']:
                # Refer to https://github.com/shaoyuanlo/EDANet/blob/master/train/class_weights_Cityscapes.PNG
                # Try to weight the loss to tell the network how to learn details
                self.build_loss(weight=Cityscape.class_weight)
            elif self.mode in ['fine_tune']:
                self.lovasz_loss()
            else:
                self.predict()

    @staticmethod
    def Upsampling(inputs, scale):
        return tf.image.resize_bilinear(inputs, size=[tf.shape(inputs)[1] * scale, tf.shape(inputs)[2] * scale])

    @staticmethod
    def _div_maybe_zero(total_loss, num_present):
        """Normalizes the total loss with the number of present pixels."""
        return tf.to_float(num_present > 0) * tf.div(total_loss,
                                                     tf.maximum(1e-5, num_present))



This is an online hard example mining code, maybe in the future I will this model open source, but not now. Hope that will give you some help.
@cena001plus

**Comment by cena001plus** [comment]
> residual


I used another open source project, Semantic-Segmentation-Suite. I think you have a reference to it. I think it is a very good open source project. The content is very good, but I don't currently support BiseNet XCeption. I want to verify this. Algorithm, see you open source this, try it out can run, but can not see the effect of algorithm training, a little regrettable. I hope that you can improve this and thank you for the open source contribution.

@pdoublerainbow
